modules/component/src/core/cd-aware/strategy.ts

```diff
@@ -0,0 +1,339 @@
+// @Notice this part of the code is in the coalescing PR https://github.com/ngrx/platform/pull/2456
+// import { generateFrames } from '../projections/generateFrames';
+// import { coalesce } from '../operators/coalesce';
+import { defer, from, MonoTypeOperatorFunction, Observable } from 'rxjs';
+import {
+  ChangeDetectorRef,
+  NgZone,
+  ɵdetectChanges as detectChanges,
+  ɵmarkDirty as markDirty,
+} from '@angular/core';
+import { hasZone, isIvy } from '../utils';
+import { mapTo } from 'rxjs/operators';
+
+/** A shared promise instance to cause a delay of one microtask */
+let resolvedPromise: Promise<void> | null = null;
+
+function getResolvedPromise() {
+  resolvedPromise = resolvedPromise || Promise.resolve();
+  return resolvedPromise;
+}
+
+function getSaveDurationSelector(ngZone: NgZone): () => Observable<number> {
+  return () => defer(getResolvedPromise).pipe(mapTo(1));
+  // @Notice this part of the code is in the coalescing PR https://github.com/ngrx/platform/pull/2456
+  /* hasZone(ngZone)
+      ? generateFrames(
+          (window as any).__zone_symbol__requestAnimationFrame,
+          (window as any).__zone_symbol__cancelAnimationFrame
+        )
+      : generateFrames();*/
+}
+
+export interface StrategyFactoryConfig {
+  component: any;
+  ngZone: NgZone;
+  cdRef: ChangeDetectorRef;
+}
+
+export interface CdStrategy<T> {
+  behaviour: (cfg?: any) => MonoTypeOperatorFunction<T>;
+  render: () => void;
+  name: string;
+}
+
+export const DEFAULT_STRATEGY_NAME = 'idle';
+
+export interface StrategySelection<U> {
+  idle: CdStrategy<U>;
+
+  [key: string]: CdStrategy<U>;
+}
+
+export function getStrategies<T>(
+  cfg: StrategyFactoryConfig
+): StrategySelection<T> {
+  return {
+    idle: createIdleStrategy<T>(cfg),
+    noop: createNoopStrategy<T>(),
+    pessimistic1: createPessimistic1Strategy<T>(cfg),
```

Comment:
@SanderElias suggest to change the numbers.

We have a scale of optimism that goes from most optimistic/performat to must pessimistic/performat behavior.
Also, the optimism assumes you know exactly what you do and your code is all correct.

**Optimism Scale**
- 1. `optimistic2`
- 2. `optimistic1`
- 3. `pessimistic2`
- 4. `pessimistic1`

Reply:
In the code it looks like:

```html
<!-- push pipe -->
{{value1$ | ngrxPush: 'level1'}}
<!-- let directive-->
<ng-container *ngrxLet="value$ as v; let w" strategy="level1">
  Value: {{v}} {{w}}
</ng-container>
```

Reply:
hey there, since I am very interested in the ngrx/component package and can see a huge benefit from those strategies, here are just my 2 cents.

Imho the current strategy API is way to complicated and unintuitive for the user. If I wouldn't know by heart what those strategy names do internally, I would have to look up the behavior pretty much every time I want to use it.

I like APIs being transparent and intuitive to use.
In a first step I would suggest to get rid of the internal `inZone` condition. A strategy should always behave the same independent of the zone. Implementing this gives users the confidence that the configured strategy will always behave the same.

Having the zone condition removed, the only difference between the strategies left is the `ChangeDetection` handling. I see 5 different approaches to implement:

- **markDirty** (or bottomUp): calls `markDirty` / `markForCheck`
- **detectChanges** (or topDown): uses `coalesce` on component scope and calls `detectChanges`
- **instant**: no coalescing, just call `detectChanges`
-**asyncLike**: mirror async pipe
-**noop**: don't do any changeDetection, just emit values

Reply:
I like the approach of ignoring the zone related thingy a bit. 
It helped me to think a bit more broad. 

Regarding **instant** as the coalescing happens as microtask over a Promise it is instant, menaing in the same tick.


There is the concept of local and global in my head...

So markForCheck and markDirty got up and render over AppRef.tick. 
DetectChanges doesn't. It only renders 'locally' from the current component down.

Also coalescing and local rendering should always happen together

**What about: **

- VE/I - Options for ViewEngine / Ivy. (* after the / means same for both )
- mFC - `cdRef.markForCheck`
- dC - `cdRef.detectChanges`
- ɵMD - `ɵmarkDirty` 
- ɵDC - `ɵdetectChanges`
- LV  - `LView`
- C - `Component`
  
| Name        | ZoneLess VE/I | Render Method VE/I  | Coalescing VE/I  |
|-------------| --------------- | ------------------------ | ----------------- | 
| `asyncLike` | ❌/❌            | mFC / mFC                 | ❌                      | 
| `global`      | ❌/✔️            | mFC  / ɵMD               | ❌                      |
| `local`         | ✔️/✔️            | dC / ɵDC                    | ✔️ + C/ LV         | 
| `noop`        | ❌/❌           | no rendering               | ❌                      |

Reply:
When looking at this today, I really like that there are only 2 real strategies. 
`global` and `local`. The 4 initial versions of strategies where based on a not existing set of test cases. As they work there is no need for this separation.

The naming should definitely be related to the rendering process.
detectChanges renders the very component down, markDirty/markForCheck the tree upwards.

`asyncLike` => `native` as it is the native behavior of Angulars `async` pipe?

Reply:
I still don't like global, local. Maybe something with component level?

Reply:
Another naming option would be "root" & "branch" instead of global/local. I like that because it is kinda simular to the general tree structure of the component architecture.

Reply:
`asyncLike` => `native` is quite cool. But why would you invent new names instead of just using `markForCheck` and `detectChanges` as names?

Reply:
Because the render methods use differ based on the platform you ran.

the concept is the same `global` or `bottom-up`, the technical implementation is different and is also related to different conditions. e.g. `global` uses `markForCheck` in ViewEngine and markDirty in ivy.

Reply:
Ok, the current naming is: `native`, `noop`, `local`, `global`.

The team has to decide what to name `global` and `local`